core/lib/zksync_core/src/fee_model.rs

```diff
@@ -0,0 +1,386 @@
+use std::{fmt, sync::Arc};
+
+use zksync_types::{
+    fee_model::{
+        BatchFeeInput, FeeModelConfig, FeeModelConfigV2, FeeParams, FeeParamsV1, FeeParamsV2,
+        L1PeggedBatchFeeModelInput, PubdataIndependentBatchFeeModelInput,
+    },
+    U256,
+};
+use zksync_utils::ceil_div_u256;
+
+use crate::l1_gas_price::L1GasPriceProvider;
+
+/// Trait responsible for providing fee info for a batch
+pub trait BatchFeeModelInputProvider: fmt::Debug + 'static + Send + Sync {
+    /// Returns the batch fee with scaling applied. This may be used to account for the fact that the L1 gas and pubdata prices may fluctuate, esp.
+    /// in API methods that should return values that are valid for some period of time after the estimation was done.
+    fn get_batch_fee_input_scaled(
+        &self,
+        l1_gas_price_scale_factor: f64,
+        l1_pubdata_price_scale_factor: f64,
+    ) -> BatchFeeInput {
+        let params = self.get_fee_model_params();
+
+        match params {
+            FeeParams::V1(params) => BatchFeeInput::L1Pegged(compute_batch_fee_model_input_v1(
+                params,
+                l1_gas_price_scale_factor,
+            )),
+            FeeParams::V2(params) => {
+                BatchFeeInput::PubdataIndependent(compute_batch_fee_model_input_v2(
+                    params,
+                    l1_gas_price_scale_factor,
+                    l1_pubdata_price_scale_factor,
+                ))
+            }
+        }
+    }
+
+    /// Returns the batch fee input as-is, i.e. without any scaling for the L1 gas and pubdata prices.
+    fn get_batch_fee_input(&self) -> BatchFeeInput {
+        self.get_batch_fee_input_scaled(1.0, 1.0)
+    }
+
+    /// Returns the fee model parameters.
+    fn get_fee_model_params(&self) -> FeeParams;
+}
+
+/// The struct that represents the batch fee input provider to be used in the main node of the server, i.e.
+/// it explicitly gets the L1 gas price from the provider and uses it to calculate the batch fee input instead of getting
+/// it from other node.
+#[derive(Debug)]
+pub(crate) struct MainNodeFeeInputProvider {
```

Comment:
I'd also suggest renaming this from 'MainNodeFeeInputProvider' to 'FeeInputProvider'.

(as we're also creating this object in external node ;-)

Reply:
@mm-zk I would personally keep it this way. While it is instantiated in the external node, it is only instantiated for the tx sender (i.e. API estimations). The fee input provider used by the state keeper of the EN is different